ci/publish-artifactory.sh

```diff
@@ -73,4 +73,10 @@ else
         done
         )
     done
+    for base in non-repudiation-core non-repudiation-client; do
```

Comment:
Is there a reason why we cannot back these up to the gcp bucket from here? It just seems to add complexity to have the assembly repo be aware of those artifacts only to push them to a gcp bucket (one more thing that is version dependent).

If we do want to go down the route of backing everything up from the assembly repo you also need the EE JSON API and EE trigger service and maybe other stuff I forgot about.

Reply:
As it stands, the split process here does not need write access to that GCS bucket. In terms of credentials management, it seems safer to add those artifacts to Artifactory than to have write access here just for them. From a pipeline design perspective, I think it's also better if we have a single place responsible for pushing all of our artifacts in one blob to our long-term backup storage, rather than having multiple repos push seemingly arbitrary subsets of what we consider the set of artifacts for a version.

I don't think this adds more version-dependency than we already have, as we already need to know pretty much the entire list of artifacts. From my perspective, version-independence in the assembly repo is no longer a credible objective.

Re: EE JSON API and EE Trigger Service, I had indeed missed those. Thanks for pointing that out! I've now double-checked all of the artifacts in that GCS bucket (for the latest non-split release), and we're also missing the daml-script and trigger-runner JARs; I've added a new commit adding those.

Reply:
I agree that there is no chance of version-independence in the assembly repo. But I think just because we cannot fully decouple things doesn’t mean we shouldn’t try to decouple things where we can. Here we have a chance to handle things for current and potential future artifacts that we add purely in the Daml repo with no coupling to the assembly repo. That seems quite desirable to me. I agree uploading from different places does create some problems but we can separate into different folders or something like that to make that work

Reply:
> a single place responsible for pushing all of our artifacts in one blob to our long-term backup storage
That definitely aligns with my vision for the assembly repo: it is the source of truth for listing what constitutes a release. (Of course the artifacts themselves are generated by the repos where development happens, but packaging them into a cohesive distribution should happen centrally, in the assembly repo.)

Reply:
(To be clear, emphasis on "vision" -- my comment is not intended to immediately impact this PR.)

Reply:
As mentioned on slack, happy to leave the decision here to @adriaanm-da and the runtime team.

Reply:
@adriaanm-da Sorry for falling behind on this discussion, I just talked to @garyverhaegen-da and @cocreature (many thanks to both for the patience in guiding me through this).

From what I understood, leaving it up to individual repos to publish their artifacts and then have the assembly depend on those as if they were external artifacts to use is simpler and more prevalent (that's the approach you would take in adding any dependency to anything, basically).

On the other hand, collecting the publishing concern in a single place is more involved but gives us a better control with regards to publishing specific artifacts, which is in this case probably something desirable enough to be favorable over simplicity (please don't ever say I said something like that).

Worst case scenario, if we see there are ways in which we can improve this, there will be chances for us to do so in the future.

I think Moritz makes a good point overall and normally I'd agree with him but I'd go Gary's way given these premises.

Reply:
Thanks for the summary, Stefano, and the good discussion, everyone. I think the path forward and the involved trade-offs are clear, as is the path forward (thanks, Gary!). Let's keep evaluating how we can improve, of course.

Reply:
One more comment from our earlier conversation: With this PR we are still a long way from actually having everything go through the assembly repo (Maven artifacts, NPM artifacts, the stuff Canton publishes are just a few examples). If we really don’t want to go down the route of only pushing the minimum set of artifacts through the assembly repo, I think we should go all the way to the other direction and not end up with some unprincipled intermediate solution where we neither benefit from decoupling as much as possible nor from having a unified location for all public artifacts.

Reply:
Yes, I missed that, thanks for bringing that up. If we want to use the `assembly` repo as the one-stop shop for all our publishing, it could make sense to go all the way in.

Reply:
Maybe I'm misunderstanding, or not expressing myself well, or both -- let me elaborate on what I said earlier:
> Of course the artifacts themselves are generated by the repos where development happens, but packaging them into a cohesive distribution should happen centrally, in the assembly repo.

By this I meant that individual artifacts are built, tested and published to whichever repo makes sense for the language they're written in. A scala jar for some component is published to maven central by the same build tool that was used to build and test that component.

By "packaging them into a cohesive distribution" I meant assembling all those individual artifacts (at the right versions) into a tarball or equivalent for distribution of our stack as a whole to our customers. The "tarball" could be as small as a json file or maven BOM that list all artifacts in a specific release, it could be configuration for daml assistant, or a windows installer.

The assembly repo governs building (and, eventually, integration testing) the Daml distribution. Individual repos own their build and CI, and should publish to a canonical location, so that their artifacts can be consumed directly between teams or even customers, if desirable.